config/config.go

```diff
@@ -0,0 +1,75 @@
+package config
+
+import (
+	"database/sql"
+	"fmt"
+	"log"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/spf13/viper"
+)
+
+type Config struct {
+	DSN      string
+	InfoLog  *log.Logger
+	ErrorLog *log.Logger
+	DB       *sql.DB
+	Addr     *string
+}
+
+// Init the application's configuration
+func Init(infoLog, errorLog *log.Logger) (*Config, error) {
+
+	//Read config file
+	viper.SetConfigName("config")
+	viper.SetConfigType("json")
+	viper.AddConfigPath(".")
+	err := viper.ReadInConfig()
+
+	if err := viper.ReadInConfig(); err != nil {
+		if _, ok := err.(viper.ConfigFileNotFoundError); ok {
+			// Config file not found
+			panic(fmt.Errorf("fatal error config file not found: %s", err))
+
+		} else {
+			// Config file was found but another error was produced
+			panic(fmt.Errorf("fatal error config file: %s", err))
+		}
+	}
+
+	// Get the server address/port
+	addr := viper.GetString("addr")
+
+	// Create the Data Source Name
+	// *We need to use the parseTime=true parameter in our
+	// DSN to force it to convert TIME and DATE fields to time.Time. Otherwise it returns these as
+	// []byte objects.
+	username := viper.GetString("db.username")
+	password := viper.GetString("db.password")
+	dbHost := viper.GetString("db.host")
+	dbPort := viper.GetString("db.port")
+	dbName := viper.GetString("db.name")
+	dsn := fmt.Sprintf("%s:%s@(%s:%s)/%s?parseTime=true", username, password, dbHost, dbPort, dbName)
+
+	// Open db connection
+	db, err := openDB(dsn)
+	if err != nil {
+		errorLog.Fatalf("message: unable to open db connection, type: database, err: %v", err)
+		return nil, err
+	}
```

Comment:
IMHO, you should only open the connection when using it, and close it after using it.

Also, I think you are mixing responsibilities here. The idea of the config is to have the configuration ready for when you are going to use it, not to initialize this kind of connection.

This (alongside moving the `OpenDB` function) should be moved/done at the repositories level, just before making the corresponding use of DB.

Reply:
According to:

https://golang.org/pkg/database/sql/#OpenDB

> the OpenDB function should be called just once. It is rarely necessary to close a DB.

And according to a book called "Let's Go!" by Alex Edwards:

> The sql.Open() function returns a sql.DB object. This isn’t a database connection — it’s a pool of many connections. Go manages these connections as needed, automatically opening and closing connections to the database via the driver.
> 
> The connection pool is intended to be long-lived. In a web application it’s normal to initialize the connection pool in your main() function and then pass the pool to your handlers. You shouldn’t call sql.Open() in a short-lived handler itself — it would be a waste of memory and network resources.

Therefore, as the connection is required to be initialized just once, at the time I was writing the code, it made sense to me to put it in a the configuration file and inject it as a dependency. However, I agree that I can move it to the repositories level as well and initialize it there as it's part of it.

What do you think?